### PR TITLE
Use supervisord instead of forever to keep datalab servers running

### PR DESCRIPTION
**Pull request by yebrahim**
We're switching to supervisor to be able to spawn multiple processes inside the docker container.

### CODE REVIEW

File: containers/gateway/content/run.sh [hunk 1]
```diff
@@ -35,11 +35,13 @@ n=0; while true; do
...
   echo "Starting Datalab Kernel Gateway in silent mode, for debug output, rerun with an additional '-e DATALAB_DEBUG=true' argument"
-  FOREVER_CMD="${FOREVER_CMD} -s"
+else
+  # The Datalab job logs to this file, make sure it exists and tail it
+  touch /var/log/gateway.stdout.log
+  tail -f /var/log/gateway.stdout.log &
 fi
 
-${FOREVER_CMD} /datalab/web/kernelproxy/app.js
+supervisord -c /etc/supervisor/conf.d/gateway.conf
```
Comment (qimingj) [inline containers/gateway/content/run.sh]: Can we do the same for jupyter command in the same file?

File: containers/base/Dockerfile [hunk 2]
```diff
@@ -27,7 +27,7 @@ RUN echo "deb-src http://ftp.us.debian.org/debian testing main" >> /etc/apt/sour
     apt-get update -y && \
     apt-get install --no-install-recommends -y -q python unzip ca-certificates build-essential \
     libatlas-base-dev liblapack-dev gfortran libpng-dev libfreetype6-dev libxft-dev libxml2-dev \
-    python-dev python-setuptools python-zmq openssh-client wget curl git pkg-config && \
+    python-dev python-setuptools python-zmq openssh-client wget curl git pkg-config supervisor && \
```
Comment (qimingj) [inline containers/base/Dockerfile]: Not sure if we need to do anything given the license here https://github.com/Supervisor/supervisor/blob/master/LICENSES.txt?

Comment (yebrahim) [inline containers/base/Dockerfile]: Need @Di-Ku input on this

File: tools/supervisor/datalab.conf [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[supervisord]
```
Comment (nikhilk) [inline tools/supervisor/datalab.conf]: I think this file should go into the containers directory.
Assumption is that this is only going to be used within the docker container, not outside.

File: tools/supervisor/gateway.conf [hunk 4]
```diff
@@ -0,0 +1,10 @@
+[supervisord]
```
Comment (nikhilk) [inline tools/supervisor/gateway.conf]: And this one too.

File: containers/base/build.sh [hunk 5]
```diff
@@ -18,12 +18,17 @@
...
+# Copy local pydatalab if supplied
 if [ -n "$1" ]; then
   rsync -avp "$1" pydatalab;
 else
   # Create empty dir to make docker build happy.
   mkdir -p pydatalab;
 fi
+
+# Copy supervisord files
+rsync -avp ../supervisor/ supervisor
```
Comment (ojarjur) [inline containers/base/build.sh]: Adding more rsync calls to the build script makes me think the files are in the wrong place.

I wouldn't put any supervisor configs in the base image, but rather put the supervisor config for the Datalab frontend in the 'datalab' image and the other configs in the 'gateway' image.

That, in turn, means we would just have those files in the 'containers/datalab/content' and 'containers/gateway/content' directories.

Comment (nikhilk) [inline containers/base/build.sh]: I agree - I also wasn't sure why all these rsync calls. Would be simpler to have them in the right place to begin with.

**Comment by yebrahim** [comment]
Closing this for now until we sort out the licensing issues.

**Pull request closed without merging**